### PR TITLE
Publishing infra V3

### PR DESCRIPTION
**Pull request by epananth**
Issue - https://github.com/dotnet/arcade/issues/5303

V3 testing -> https://dnceng.visualstudio.com/internal/_build/results?buildId=757003&view=results
Promote -> https://dnceng.visualstudio.com/internal/_build/results?buildId=757112&view=results

v2 testing -> https://dnceng.visualstudio.com/internal/_build/results?buildId=743716&view=results

### CODE REVIEW

File: NuGet.config [hunk 1]
```diff
@@ -8,7 +8,6 @@
     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
     <add key="dotnet-tools" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-tools/nuget/v3/index.json" />
     <add key="dotnet-eng" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json" />
-    <add key="vs-buildservices" value="https://pkgs.dev.azure.com/azure-public/vside/_packaging/vs-buildservices/nuget/v3/index.json" />
```
Comment (riarenas) [inline NuGet.config:11]: I don't think this PR should remove this. probably a bad merge.

File: azure-pipelines.yml [hunk 2]
```diff
@@ -213,14 +213,13 @@ stages:
...
-      # once this issue is resolved: https://github.com/dotnet/arcade/issues/2871
-      enableSymbolValidation: false
+      publishingInfraVersion: 3
+
       # Sourcelink validation isn't passing for Arcade due to some regressions. This should be
       # enabled back once this issue is resolved: https://github.com/dotnet/arcade/issues/2912
       enableSourceLinkValidation: false
       publishDependsOn:
-      - Validate
+      - Validate
```
Comment (riarenas) [inline azure-pipelines.yml:222]:
```suggestion
      - Validate
```

File: eng/common/post-build/publish-using-darc.ps1 [hunk 3]
```diff
@@ -47,9 +48,10 @@ try {
       $optionalParams.Add($SigningValidationAdditionalParameters) | Out-Null
     }
   }
-
+
```
Comment (riarenas) [inline eng/common/post-build/publish-using-darc.ps1:51]: [nit] extra spaces

File: eng/common/post-build/publish-using-darc.ps1 [hunk 4]
```diff
@@ -47,9 +48,10 @@ try {
       $optionalParams.Add($SigningValidationAdditionalParameters) | Out-Null
     }
   }
-
+  
   & darc add-build-to-channel `
 	--id $buildId `
+  --publishing-infra-version $PublishingInfraVersion `
```
Comment (riarenas) [inline eng/common/post-build/publish-using-darc.ps1:54]: [nit] wrong indentation

File: eng/promote-build.yml [hunk 5]
```diff
@@ -1,19 +1,57 @@
...
-#   - EnableSDLValidation                           ; Whether SDL validation should be performed
-#   -   SDLValidationContinueOnError                ; Whether SDL validation errors should be ignored
-#   -   SDLValidationCustomParams                   ; Custom parameters for SDL validation
-#   - PublishInstallersAndChecksums                 ; Whether installers and checksums should be published
-#   - SymbolPublishingAdditionalParameters          ; Additional (MSBuild) properties to be passed to symbol publishing
-#   - ArtifactsPublishingAdditionalParameters       ; Additional (MSBuild) properties to be passed to asset publishing
+
+parameters:
+  - name: PublishingInfraVersion
+    displayName: Which version of the should be used?
```
Comment (riarenas) [inline eng/promote-build.yml:5]: unfinished sentence

File: eng/publishing/v3/publish-assets.yml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      displayName: 'Authenticate to AzDO Feeds'
+
+    - task: PowerShell@2
+      displayName: Enable cross-org publishing
+      inputs:
+        filePath: $(Build.SourcesDirectory)/eng/common/enable-cross-org-publishing.ps1
+        arguments: -token $(dn-bot-dnceng-artifact-feeds-rw)
+
+    - task: PowerShell@2
+      displayName: Packages & Blobs Publishing
```
Comment (riarenas) [inline eng/publishing/v3/publish-assets.yml:55]:
```suggestion
      displayName: Publish packages & blobs
```

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/GenerateBuildManifest.proj [hunk 7]
```diff
@@ -40,6 +40,7 @@
                            BuildData="$(ManifestBuildData)"
                            RepoUri="$(BUILD_REPOSITORY_URI)"
                            RepoBranch="$(BUILD_SOURCEBRANCH)"
-                           RepoCommit="$(BUILD_SOURCEVERSION)" />
+                           RepoCommit="$(BUILD_SOURCEVERSION)"
+                           AzureDevOpsCollectionUri="testUri"/>
```
Comment (riarenas) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/GenerateBuildManifest.proj:44]: Probably a leftover debugging change that should be removed?

File: eng/publishing/v3/signing-validation.yml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    - task: NuGetAuthenticate@0
+      displayName: 'Authenticate to AzDO Feeds'
+
+    - task: PowerShell@2
+      displayName: Enable cross-org publishing
+      inputs:
+        filePath: $(Build.SourcesDirectory)/eng/common/enable-cross-org-publishing.ps1
+        arguments: -token $(dn-bot-dnceng-artifact-feeds-rw)
+
+    # Signing validation will optionally work with the buildmanifest file which is downloaded from
```
Comment (riarenas) [inline eng/publishing/v3/signing-validation.yml:49]: I don't understand this comment. This template shouldn't be modified, so how do I make it work with the manifest?

File: azure-pipelines.yml [hunk 9]
```diff
@@ -213,14 +213,13 @@ stages:
 
   - template: eng\common\templates\post-build\post-build.yml
     parameters:
-      # Symbol validation isn't being very reliable lately. This should be enabled back
```
Comment (riarenas) [inline azure-pipelines.yml:216]: Are you removing this because we should re-enable symbol validation? (I do see that 2871 is closed)

**Comment by riarenas** [comment]
Do you have a full example of v3 publishing? the one provided in the PR description didn't do much since it looks like the build was already in the channel

```
Starting: Publish Using Darc
==============================================================================
Task         : PowerShell
Description  : Run a PowerShell script on Linux, macOS, or Windows
Version      : 2.170.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
==============================================================================
Generating script.
Formatted command: . 'D:\a\1\s\eng\common\post-build\publish-using-darc.ps1' -BuildId 59272 -PublishingInfraVersion 3 -AzdoToken '***' -MaestroToken '***' -WaitPublishingFinish true -EnableSourceLinkValidation false -EnableSigningValidation false -EnableNugetValidation false -PublishInstallersAndChecksums false -ArtifactsPublishingAdditionalParameters '' -SigningValidationAdditionalParameters ''
========================== Starting Command Output ===========================
"C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\719a9594-cd4b-4dd8-aa0d-7413c43ce570.ps1'"
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-preview.6.20310.4/dotnet-sdk-5.0.100-preview.6.20310.4-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-preview.6.20310.4/dotnet-sdk-5.0.100-preview.6.20310.4-win-x64.zip
dotnet-install: Adding to current process PATH: "D:\a\1\s\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
Installing Darc CLI version 1.1.0-beta.20379.1...
You may need to restart your command window if this is the first dotnet tool you have installed.
'D:\a\1\s\.dotnet\dotnet.exe' tool install microsoft.dotnet.darc --version 1.1.0-beta.20379.1 --add-source 'https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json' -v minimal -g
  Determining projects to restore...
  Restored C:\Users\VssAdministrator\AppData\Local\Temp\atnbr1m1.0en\restore.csproj (in 4.17 sec).
You can invoke the tool using the following command: darc
Tool 'microsoft.dotnet.darc' (version '1.1.0-beta.20379.1') was successfully installed.
Build '59272' is already on all target channel(s).
done.
Finishing: Publish Using Darc
```

File: eng/common/templates/post-build/setup-maestro-vars.yml [hunk 10]
```diff
@@ -1,34 +1,34 @@
...
+          buildType: current
+          artifactName: ReleaseConfigs
 
     - task: PowerShell@2
       name: setReleaseVars
       displayName: Set Release Configs Vars
       inputs:
         targetType: inline
         script: |
+          dir Env:
```
Comment (riarenas) [inline eng/common/templates/post-build/setup-maestro-vars.yml:28]: debug statement?

File: eng/promote-build.yml [hunk 11]
```diff
@@ -92,25 +130,41 @@ stages:
...
-- template: \eng\common\templates\post-build\post-build.yml
-  parameters:
-    enableSourceLinkValidation: variables['EnableSourceLinkValidation']
-    enableNugetValidation: variables['EnableNugetValidation']
-    enableSigningValidation: variables['EnableSigningValidation']
-    
-    validateDependsOn:
-      - prepare_promotion
+- ${{ if le(parameters.PublishingInfraVersion, 2) }}:
+  - template: \eng\common\templates\post-build\post-build.yml
```
Comment (riarenas) [inline eng/promote-build.yml:137]: It's so frustrating that the only thing different in these two blocks is the template to use, but I'm pretty sure the yaml parser won't let us factor just this into a variable.

File: NuGet.config (hunk 1 above)
Comment (mmitche) [reply NuGet.config:11]: What do we use this for?

Comment (riarenas) [reply NuGet.config:11]: Was introduced by https://github.com/dotnet/arcade/commit/f4db815522a698d47894fdb595fbab1b901bf90a so I'm guessing some of the packages for rich code navigation are hosted there.

File: eng/common/templates/post-build/channels/generic-internal-channel.yml [hunk 12]
```diff
@@ -138,6 +143,7 @@ stages:
         inputs:
           filePath: eng\common\sdk-task.ps1
           arguments: -task PublishArtifactsInManifest -restore -msbuildEngine dotnet
+            /p:publishingInfraVersion=2
```
Comment (mmitche) [inline eng/common/templates/post-build/channels/generic-internal-channel.yml:146]: Nit: casing

File: eng/common/templates/post-build/channels/generic-public-channel.yml [hunk 13]
```diff
@@ -137,6 +142,7 @@ stages:
         inputs:
           filePath: eng\common\sdk-task.ps1
           arguments: -task PublishArtifactsInManifest -restore -msbuildEngine dotnet
+            /p:publishingInfraVersion=2
```
Comment (mmitche) [inline eng/common/templates/post-build/channels/generic-public-channel.yml:145]: Nit: casing

File: eng/promote-build.yml [hunk 14]
```diff
@@ -1,19 +1,57 @@
...
-#   - SymbolPublishingAdditionalParameters          ; Additional (MSBuild) properties to be passed to symbol publishing
-#   - ArtifactsPublishingAdditionalParameters       ; Additional (MSBuild) properties to be passed to asset publishing
+
+parameters:
+  - name: PublishingInfraVersion
+    displayName: Which version of the should be used?
+    type: number
+    default: 2
+    values:
+      - 2
```
Comment (mmitche) [inline eng/promote-build.yml:9]: This allows for restriction of potential input values?

File: eng/promote-build.yml [hunk 15]
```diff
@@ -1,19 +1,57 @@
...
+
+  - name: ArtifactsPublishingAdditionalParameters
+    displayName: Additional (MSBuild) properties for general asset publishing
+    type: string 
+    default: ' '
+    
+  - name: SigningValidationAdditionalParameters
+    displayName: Additional (MSBuild) properties for signing validation
+    type: string 
+    default: ' '
```
Comment (riarenas) [inline eng/promote-build.yml:55]: This new YAML is missing the SDL validation parameters. Is that intentional?

**Comment by epananth** [comment]
Finally got the end to end test working -> https://dnceng.visualstudio.com/internal/_build/results?buildId=771946&view=results

File: eng/promote-build.yml (hunk 15 above)
Comment (epananth) [reply eng/promote-build.yml:55]: as discussed moved it back in.

File: NuGet.config (hunk 1 above)
Comment (epananth) [reply NuGet.config]: reverted this

File: eng/promote-build.yml (hunk 11 above)
Comment (epananth) [reply eng/promote-build.yml:137]: :( yeah

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/GenerateBuildManifest.proj [hunk 16]
```diff
@@ -43,6 +43,6 @@
                            BuildData="$(ManifestBuildData)"
                            RepoUri="$(BUILD_REPOSITORY_URI)"
                            RepoBranch="$(BUILD_SOURCEBRANCH)"
-                           RepoCommit="$(BUILD_SOURCEVERSION)" />
+                           RepoCommit="$(BUILD_SOURCEVERSION)"
```
Comment (riarenas) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/GenerateBuildManifest.proj:46]: Isn't this closing tag required?

File: azure-pipelines.yml (hunk 9 above)
Comment (epananth) [reply azure-pipelines.yml:216]: I think this is taken care here -> https://github.com/dotnet/arcade/issues/3735

Comment (riarenas) [reply azure-pipelines.yml:216]: Yeah that's fine with me. We just don't want to run it during the build then?

File: eng/common/post-build/publish-using-darc.ps1 [hunk 17]
```diff
@@ -49,12 +50,13 @@ try {
...
-	--id $buildId `
-	--default-channels `
-	--source-branch master `
-	--azdev-pat $AzdoToken `
-	--bar-uri $MaestroApiEndPoint `
-	--password $MaestroToken `
+  --id $buildId `
+  --publishing-infra-version $PublishingInfraVersion `
+  --default-channels `
+  --source-branch master `
```
Comment (riarenas) [inline eng/common/post-build/publish-using-darc.ps1:56]: Should this be the "publish" branch or is that coming in a later PR?

Comment (epananth) [reply eng/common/post-build/publish-using-darc.ps1:56]: for now we are going to use the master branch itself, I used publish branch to test the changes. I have another issue for using published branch -> https://github.com/dotnet/arcade/issues/5308, I will work on that after this pr merge..

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/GenerateBuildManifest.proj [hunk 18]
```diff
@@ -43,6 +43,6 @@
                            BuildData="$(ManifestBuildData)"
                            RepoUri="$(BUILD_REPOSITORY_URI)"
                            RepoBranch="$(BUILD_SOURCEBRANCH)"
-                           RepoCommit="$(BUILD_SOURCEVERSION)" />
+                           RepoCommit="$(BUILD_SOURCEVERSION)"/>
```
Comment (riarenas) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/GenerateBuildManifest.proj:46]:
```suggestion
                           RepoCommit="$(BUILD_SOURCEVERSION)" />
```

File: azure-pipelines.yml (hunk 9 above)
Comment (epananth) [reply azure-pipelines.yml:216]: Yes,

**Pull request merged**